src/ds/address.h

```diff
@@ -39,4 +39,32 @@ namespace snmalloc
   {
     return reinterpret_cast<T*>(address);
   }
-} // namespace snmalloc
+
+  template<size_t granule, typename T = void>
+  inline T* pointer_align_down(void* p)
+  {
+    static_assert((granule & (granule - 1)) == 0);
+#if defined(__has_builtin) && __has_builtin(__builtin_align_down)
+    return reinterpret_cast<T*>(__builtin_align_down(p, granule));
+#else
+    /* XREF bits::align_down; can't use here for cyclic deps */
```

Comment:
If we move `bits::is_aligned_block` then we could move `bits.h` before `address.h`?

Reply:
`bits::hash` also depends on `address_cast`.

Reply:
I see that `bits::hash` is no more; where would you like `bits::is_aligned_block` instead?

Reply:
I think it makes most sense in `address.h`. Thanks

Reply:
Then you can use `bits.h` align functions when the compiler does not support `__builtin_align_up`.